File: RayTracer/RayTracer/GeometricObjects/Triangle.fs
namespace RayTracer.GeometricObjects

open RayTracer.BaseTypes

type Triangle(a : Vec3, b : Vec3, c : Vec3) =
    let n = Vec3Ops.Normalize (Vec3Ops.Cross (b - a) (c - a))
    interface GeometricObject with
        member this.Hit(ray : Ray) : HitRecord option =
            let temp = (Vec3Ops.Dot (a - ray.Origin) n) / (Vec3Ops.Dot ray.Direction n)

            match temp > kEpsilon with
            | true -> 
                let p = ray.GetPoint(temp)
                let n = Vec3Ops.Cross (b - a) (c - a)
                let denom = Vec3Ops.Dot n n
                let l1 = (Vec3Ops.Dot n (Vec3Ops.Cross (b - a) (p - a))) / denom
                let l2 = (Vec3Ops.Dot n (Vec3Ops.Cross (c - b) (p - b))) / denom
                let l3 = 1.0 - l1 - l2
                match 0.0 <= l1 && l1 <= 1.0 && 0.0 <= l2 && l2 <= 1.0 && 0.0 <= l3 && l3 <= 1.0 with
                | true ->
                    Some({ Normal = Vec3Ops.Normalize n
                           Position = p
                           RayT = temp })
                | false -> None
            | false -> None

File: Source/Core/GeometricObject/Basic/Triangle.cpp
#include "Triangle.h"

void Triangle::Triangle::UpdateNormal()
{
    n = (b - a) ^ (c - a);
}

void Triangle::UpdateBoundingBox()
{
    boundingBox = BBox(a, a);
    boundingBox.Merge(BBox(b, b));
    boundingBox.Merge(BBox(c, c));
}

Triangle::Triangle(Vec3D vertexA, Vec3D vertexB, Vec3D vertexC)
    : a(vertexA), b(vertexB), c(vertexC)
{
    UpdateNormal();
    UpdateBoundingBox();
}

void Triangle::setA(Vec3D vertexA) { a = vertexA; UpdateNormal(); UpdateBoundingBox(); }
void Triangle::setB(Vec3D vertexB) { b = vertexB; UpdateNormal(); UpdateBoundingBox(); }
void Triangle::setC(Vec3D vertexC) { c = vertexC; UpdateNormal(); UpdateBoundingBox(); }
Vec3D Triangle::getA() { return a; }
Vec3D Triangle::getB() { return b; }
Vec3D Triangle::getC() { return c; }
Vec3D Triangle::getNormal() { return n.normalised(); }

HitRecord Triangle::Hit(const Ray& ray)
{
    HitRecord record;
    auto temp = ((a - ray.Origin) * n) / (ray.Direction * n);
    if (temp > kEpsilon)
    {
        auto p = ray.GetPoint(temp);
        auto denom = n * n;
        auto l1 = (n * ((b - a) ^ (p - a))) / denom;
        auto l2 = (n * ((c - b) ^ (p - b))) / denom;
        auto l3 = 1.0 - l1 - l2;
        if (l1 >= 0.0 && l1 <= 1.0 && l2 >= 0.0 && l2 <= 1.0 && l3 >= 0.0 && l3 <= 1.0)
        {
            record.Hit = true;
            record.Normal = (n * -ray.Direction > 0.0 ? n.normalised() : -n.normalised());
            record.NormalFlipped = n * ray.Direction > 0.0;
            record.HitPoint = p;
            record.MaterialPtr = materialPtr;
            record.Ray = ray;
            record.T = temp;
        }
    }
    return record;
}
